### PR TITLE
[CBRD-22106] resource tracking

### PR DESCRIPTION
**Pull request by ribram**
http://jira.cubrid.org/browse/CBRD-22106

Migrate resource tracking from thread.h/thread.c. This is the final functionality to migrate from legacy thread unit.

There were four types of resources tracked:

  1. allocated pointers using private allocator
  2. fixed pages
  3. query list files
  4. critical section usage.

resources_tracker is used for first two. resource_tracker can trace resource usage and detect issues like leaks, over-usage, invalid free.

for query lists, just a counter is enough actually. leaks are never checkes, it only verifies the number of list files created within scope matches expectations.

for critical section, a different kind of tracking was required. a special csect_tracker was created to make all legacy checks.

Other changes:

  - added a fileline_location structure to store file and line.
  - added thread_entry to signature of private allocation functions.
  - create unit test for resource_tracker and csect_tracker.

### CODE REVIEW

File: src/thread/critical_section_tracker.hpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+/*
+ * critical_section_tracker.hpp - interface to track and debug critical sections usage
+ */
+
+#ifndef _CUBRID_CRITICAL_SECTION_TRACKER_HPP_
+#define _CUBRID_CRITICAL_SECTION_TRACKER_HPP_
+
+#include "critical_section.h"
+
+#include <cinttypes>
```
Comment (valerinistor) [inline src/thread/critical_section_tracker.hpp]: Is cinttypes header used? seems like not

Comment (ribram) [reply src/thread/critical_section_tracker.hpp]: no, you're right. I will remove it.

File: src/base/fileline_location.hpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    {
+      return "%s:%d";
+    }
+
+    void set (const char *fn_arg, int l_arg);
+
+    char m_file[MAX_FILENAME_SIZE];
+    int m_line;
+  };
+#define FILELINE_LOCATION_AS_ARGS(fileline) (fileline).m_file, (fileline).m_line
```
Comment (valerinistor) [inline src/base/fileline_location.hpp]: FILELINE_LOCATION_AS_ARGS macro is not used as well as print_format function

File: src/thread/thread_entry.cpp [hunk 3]
```diff
@@ -328,6 +370,46 @@ namespace cubthread
...
+  entry::end_resource_tracks (void)
+  {
+    if (!ENABLE_TRACKERS)
+      {
+	// all trackers are activated by this flag
+	return;
+      }
+    m_alloc_tracker.clear_all ();
+    m_pgbuf_tracker.clear_all ();
+    m_csect_tracker.clear_all ();
```
Comment (valerinistor) [inline src/thread/thread_entry.cpp]: Shouldn't m_csect_tracker be stopped as well after clear?

File: src/thread/thread_entry.cpp [hunk 4]
```diff
@@ -42,8 +44,43 @@
...
+  const std::size_t ALLOC_TRACK_MAX_ITEMS = 32767;
+
+  // page buffer
+  const char *PGBUF_TRACK_NAME = "Page Buffer";
+  const char *PGBUF_TRACK_RES_NAME = "pgptr";
+  const std::size_t PGBUF_TRACK_MAX_ITEMS = 1024;
+  const unsigned PGBUF_TRACK_MAX_AMOUNT = 16;       // re-fix is possible... how many to accept is debatable
+
+  // critical section
+  const char *CSECT_TRACK_NAME = "Critical Section";
```
[the remark this answers is not in the document]
Comment (ribram) [reply src/thread/thread_entry.cpp]: no, it is leaked code.

File: src/base/fileline_location.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    : m_file {}
+    , m_line (0)
+  {
+    set (fn_arg, l_arg);
+  }
+
+  void
+  fileline_location::set (const char *fn_arg, int l_arg)
+  {
+    // find filename from full path; get last charachter of '\\' or '/'
```
[the remark this answers is not in the document]
Comment (ribram) [reply src/base/fileline_location.cpp]: fixed

File: src/thread/thread_entry.cpp (hunk 3 above)
Comment (ribram) [reply src/thread/thread_entry.cpp]: fixed by resetting m_start_count

**Pull request merged**